src/typescript-worker/util.ts

```diff
@@ -4,41 +4,41 @@
  * SPDX-License-Identifier: BSD-3-Clause
  */
 
-type IteratorResultWithNext<T> = Promise<{
-  value: IteratorResult<T>;
-  next: () => IteratorResultWithNext<T>;
-}>;
-
 /**
  * Merges multiple async iterables into one iterable. Iterator next promises are
  * raced, so order is not preserved. Additional iterables can be added before or
  * during iteration.
  */
 export class MergedAsyncIterables<T> {
-  private _iterators = new Map<
-    () => IteratorResultWithNext<T>,
-    IteratorResultWithNext<T>
-  >();
-
-  add(iterable: AsyncIterable<T>) {
-    const iterator = iterable[Symbol.asyncIterator]();
-    const next = async () => ({
-      value: await iterator.next(),
-      next,
-    });
-    this._iterators.set(next, next());
-  }
+  private readonly _buffer: Array<{value: T; emitted: () => void}> = [];
+  private _numSources = 0;
+  private _notify?: () => void;
 
   async *[Symbol.asyncIterator]() {
-    while (this._iterators.size > 0) {
-      const {value, next} = await Promise.race(this._iterators.values());
-      if (value.done) {
-        this._iterators.delete(next);
-      } else {
-        yield value.value;
-        this._iterators.set(next, next());
+    while (this._numSources > 0) {
```

Comment:
There is a race condition here where if existing iterators are done, adding an iterator will not do anything. I'd try to add an error in that case below, and either make this class take a required initial iterator in the constructor (just call `this.add()`) or wait for at least one value by changing this to a do/while loop so the `await` kicks in.

Reply:
I like the error if you try to add after iteration is complete, done!

Not sure the constructor idea could work, because then you can't do this (which is what I'm doing):

```ts
    const output = new MergedAsyncIterables<BuildOutput>();
    output.add(this._handleProjectFiles(results, output));
```

Also not sure about making it always wait for an initial value, because it feels illogical to me that this pattern would never resolve. This is actually one of the existing unit tests. It seems more logical to me that if you iterate over a merged set of no iterables, you successfully get no values, rather than a deadlock.

```ts
    const merged = new MergedAsyncIterables();
    for await (const v of merged) {
    }
```

Reply:
Sounds good!